File: info/1. Алгоритмы/2. Сортировка/20. Пирамидальная сортировка/code sample/Python/sample.py
arr = [5, 0, -2, 7, 3]


def sift_up(arr, i):
    while i > 0:
        j = (i-1)//2
        if arr[i] > arr[j]:
            arr[i], arr[j] = arr[j], arr[i]
        else:
            break
        i = j


def sift_down(arr, i, last_index):
    while True:
        left_j = 2 * i + 1
        right_j = 2 * i + 2
        j = i
        if left_j < last_index and arr[left_j] > arr[j]:
            j = left_j
        if right_j < last_index and arr[right_j] > arr[j]:
            j = right_j
        if j != i:
            arr[i], arr[j] = arr[j], arr[i]
            i = j
        else:
            break


def heapify(arr):
    for i in range(len(arr)):
        sift_up(arr, i)


def heap_sort(arr):
    heapify(arr)
    last_index = len(arr) - 1
    while last_index > 0:
        arr[0], arr[last_index] = arr[last_index], arr[0]
        sift_down(arr, 0, last_index)
        last_index -= 1

print(arr)
print()
heap_sort(arr)
print(arr)

File: info/1. Алгоритмы/2. Сортировка/20. Пирамидальная сортировка/code sample/Fortran/sample.f90
program sample
implicit none

integer::array(5)

array = [5, 0, -2, 7, 3]

write(*,*) array

call heap_sort(array)

write(*,*) array



contains

function leafSearch(arr, i, last_index)
    implicit none
    integer,intent(in)::arr(:)
    integer,intent(in)::last_index
    integer,intent(in)::i
    integer::j
    integer::leafSearch
    j = i
    do
        if(2 * j + 1 >= last_index) then
            exit
        end if
        if (arr(2 * j) > arr (2 * j + 1)) then
            j = 2 * j
        else
            j = 2 * j + 1
        end if
    end do
    if (2 * j < last_index) then
        j = 2 * j
    end if
    leafSearch = j    
end function

subroutine silf_down(arr,i, last_index)
    implicit none
    integer, intent(inout)::arr(:)
    integer, intent(in)::i, last_index
    integer::j
    j = leafSearch(arr, i, last_index)
    do 
        if(arr(i) <= arr(j)) then
            exit
        end if
        j = j / 2
    end do
    
    do
        if(j <= i) then
            exit
        end if
        call swap(arr(i), arr(j))
        j=j/2
    end do

end subroutine silf_down

subroutine swap(first_val, second_val)
    implicit none
    integer, intent(inout)::first_val, second_val
    integer::temp
    temp = first_val
    first_val = second_val
    second_val = temp
end subroutine swap


subroutine heap_sort(arr)
    implicit none
    integer,intent(inout)::arr(:)
    integer::i, last_index, n
    n = size(arr, dim = 1)
    last_index = n + 1
    do i = n/2,1,-1
        call silf_down(arr,i, last_index)
    end do
    last_index = n
    do
        if(last_index == 1) then
            exit
        end if
        call swap(arr(1),arr(last_index))
        call silf_down(arr,1, last_index)
        last_index = last_index - 1
    end do    
end subroutine heap_sort

end program sample
